waspc/test/AppSpec/ValidTest.hs

```diff
@@ -323,6 +330,120 @@ spec_AppSpecValid = do
         it "returns an error if the Dummy email sender is used when building the app" $ do
           ASV.validateAppSpec (makeSpec (Just dummyEmailSender) True)
             `shouldBe` [ASV.GenericValidationError "app.emailSender must not be set to Dummy when building for production."]
+
+    -- NOTE: we need to test all types of declarations here, make sure to add
+    -- tests for new declaration types as they are validated in Valid.hs
+    describe "duplicate declarations validation" $ do
+      -- Page
+      let pageDecl = makeBasicPageDecl "testPage"
+
+      -- Route
+      let routeDecl = makeBasicRouteDecl "testRoute" "testPage"
+
+      -- Action
+      let actionDecl = makeBasicActionDecl "testAction"
+
+      -- Query
+      let queryDecl = makeBasicQueryDecl "testQuery"
+
+      -- Api
+      let apiDecl1 = makeBasicApiDecl "testApi" (AS.Api.GET, "/foo/bar")
+      -- Using a different route not to trigger duplicate route errors
+      let apiDecl2 = makeBasicApiDecl "testApi" (AS.Api.GET, "/different/route")
+
+      -- ApiNamespace
+      let apiNamespaceDecl1 = makeBasicApiNamespaceDecl "testApiNamespace" "/foo"
+      -- Using a different path not to trigger duplicate route errors
+      let apiNamespaceDecl2 = makeBasicApiNamespaceDecl "testApiNamespace" "/different/path"
+
+      -- Crud
+      let crudDecl = makeBasicCrudDecl "testCrud" "TestEntity"
+
+      -- Entity
+      let entityDecl = makeBasicEntityDecl "TestEntity"
+
+      -- Job
+      let jobDecl = makeBasicJobDecl "testJob"
+
+      it "returns no error if there are no duplicate declarations" $ do
```

Comment:
Not bad, but a lot of repetition, and it makes it harder to maintain these tests.

Same as normal code, tests benefit from clean code, as they are easier to maintain.
Although it is usually ok to err on the side of boilerplate a bit more.

Here, what I would do is write a function that generates the `it ...` thing for you. Call it something like `test ...`. You can see this pattern in some of the other tests we write.
So your tests could look like:
```hs
testErrorOnDuplicates "pages" pageDecl
testErrorOnDuplicates "routes" routeDecl
...
Or you can give it multiple declaraitons, up to you, but the point is, you can make these one line and much easier to quickly scan and maintain.
```

Reply:
I now see you are also checking error message -> you can also pass that as an argument, whole or partial.